### PR TITLE
Fix Pre-Select Semester Render Problem

### PR DESCRIPTION
**Pull request by NicholasHallman**
Functional 
- Fix the All Selected filter message when the dashboard loads for the first time.
- Fix the deselect semester bug
- Properly initialize the selected list on first load

### CODE REVIEW

File: components/semester-filter.js [hunk 1]
```diff
@@ -21,7 +21,8 @@ class SemesterFilter extends Localizer(LitElement) {
 			preSelected: { type: Array, attribute: false },
 			_filterData: { type: Array, attribute: false },
 			_bookmark: { type: String, attribute: false },
-			_search: { types: String, attribute: false }
+			_search: { type: String, attribute: false },
+			isLoading: { type: Boolean, attribute: 'is-loading' }
```
Comment (johngwilkinson) [inline components/semester-filter.js:25]: attribute name should be `loading`

File: engagement-dashboard.js [hunk 2]
```diff
@@ -346,6 +346,7 @@ class EngagementDashboard extends Localizer(MobxLitElement) {
 		<d2l-insights-semester-filter
 			page-size="10000"
 			?demo="${this.isDemo}"
+			?is-default-view="${this._isLoading}"
```
Comment (johngwilkinson) [inline engagement-dashboard.js:349]: shouldn't this be `?loading="${this._isLoading}"`?

**Review by johngwilkinson — COMMENTED** [review]
Did you test on draco or the demo page?

**Comment by NicholasHallman** [comment]
> Did you test on draco or the demo page?
Did a test on draco stack, still had the issue. latest commit solves the problem

File: components/semester-filter.js [hunk 3]
```diff
@@ -77,15 +77,21 @@ class SemesterFilter extends Localizer(LitElement) {
...
+				const setSelected = new Set(this.selected);
+				setSelected.add(e.detail.itemId);
+				this.selected.clear();
+				[...setSelected].forEach(s => this.selected.push(s));
 			} else {
-				this.selected.delete(e.detail.itemId);
+				const setSelected = new Set(this.selected);
+				setSelected.delete(e.detail.itemId);
+				this.selected.clear();
+				[...setSelected].forEach(s => this.selected.push(s));
```
Comment (NicholasHallman) [inline components/semester-filter.js:88]: This code modifies the data.serverData.selectedSemesterIds in place so as not to replace the reference with a new array

File: engagement-dashboard.js [hunk 4]
```diff
@@ -687,7 +688,7 @@ class EngagementDashboard extends Localizer(MobxLitElement) {
 	_semesterFilterChange(event) {
 		event.stopPropagation();
 		filterEventQueue.add(this.localize('alert:updatedFilter', { chartName: this.localize('semesterFilter:name') }));
-		this._serverData.selectedSemesterIds = event.detail;
+		this._serverData.selectedSemesterIds = [...this._serverData.selectedSemesterIds];
```
Comment (NicholasHallman) [inline engagement-dashboard.js:691]: the setter calls reload so once the semester filter is done modifying the array we set it here to trigger the api call

**Review by johngwilkinson — APPROVED** [review]
Have tested, looks good!

(Found an unrelated, pre-existing bug where urlState changes won't cause data to load even if neeed)

File: components/semester-filter.js [hunk 5]
```diff
@@ -77,15 +77,21 @@ class SemesterFilter extends Localizer(LitElement) {
...
+				const setSelected = new Set(this.selected);
+				setSelected.add(e.detail.itemId);
+				this.selected.splice(0);
+				[...setSelected].forEach(s => this.selected.push(s));
 			} else {
-				this.selected.delete(e.detail.itemId);
+				const setSelected = new Set(this.selected);
+				setSelected.delete(e.detail.itemId);
+				this.selected.splice(0);
+				[...setSelected].forEach(s => this.selected.push(s));
```
Comment (bemailloux) [inline components/semester-filter.js:88]: oh interesting - is this modifying all the elements in the array while avoiding reassignment (for mobx reasons I guess)?

**Review by bemailloux — APPROVED** [review]
LGTM!

File: components/semester-filter.js (hunk 5 above)
Comment (NicholasHallman) [reply components/semester-filter.js:88]: Yep!

Comment (NicholasHallman) [reply components/semester-filter.js:88]: Also, do you mind pulling the branch and trying this on draco stack? John and I would like one more eyeball to make sure there isn't something we missed

Comment (bemailloux) [reply components/semester-filter.js:88]: Ok, I've completed my testing and I think we're good to go. One edge-case bug gave me the run around for a bit, but it requires a semester with no children to be selected as default, plus org unit truncation, for the bug to manifest. That's probably not an uncommon occurrence but it's also not a blocker for this PR imo.

(ref: https://d2l.slack.com/archives/C031HN77DSA/p1645718565088819)

**Comment by johngwilkinson** [comment]
Sauce failed. I'll rerun it

**Comment by johngwilkinson** [comment]
> Sauce failed. I'll rerun it

@NicholasHallman in the meantime, you might want to check if the tests pass locally for you - it occurs to me these changes could potentially break a test or two

**Review by johngwilkinson — APPROVED** [review]

**Pull request merged**